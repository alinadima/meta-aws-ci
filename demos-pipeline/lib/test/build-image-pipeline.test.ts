import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import {
  BuildImagePipelineStack,
  ImageKind,
} from "../lib/build-image-pipeline";
import { Repository } from "aws-cdk-lib/aws-ecr";
import { Bucket } from "aws-cdk-lib/aws-s3";

describe("Build Image Pipeline", () => {
  const env = { account: "111111111111", region: "eu-central-1" };

  test("Build Image Pipeline Instantiates", () => {
    const app = new cdk.App();
    const repoStack = new cdk.Stack(app, "RepoStack", { env });
    const repository = new Repository(repoStack, "Repository", {});
    const dataBucket = new Bucket(repoStack, "Bucket", {});

    const props = {
      env,
      imageKind: ImageKind.Ubuntu22_04,
      repository,
      dataBucket,
    };

    const stack = new BuildImagePipelineStack(repoStack, "MyTestStack", props);
    const template = Template.fromStack(stack);
    template.resourceCountIs("AWS::CodePipeline::Pipeline", 1);
  });

  test("Snapshot", () => {
    const app = new cdk.App();
    const repoStack = new cdk.Stack(app, "RepoStack", { env });
    const repository = new Repository(repoStack, "Repository", {});
    const dataBucket = new Bucket(repoStack, "Bucket", {});

    const props = {
      env,
      imageKind: ImageKind.Ubuntu22_04,
      repository,
      dataBucket,
    };

    const stack = new BuildImagePipelineStack(repoStack, "MyTestStack", props);
    const template = Template.fromStack(stack);
    expect(template).toMatchSnapshot();
  });
});
