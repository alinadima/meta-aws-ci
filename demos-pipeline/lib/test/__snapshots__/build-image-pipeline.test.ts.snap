// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Build Image Pipeline Snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "BuildImagePipeline01DF392E": {
      "DependsOn": [
        "BuildImagePipelineRoleDefaultPolicy211D9F92",
        "BuildImagePipelineRole576404EF",
      ],
      "Properties": {
        "ArtifactStore": {
          "EncryptionKey": {
            "Id": {
              "Fn::GetAtt": [
                "BuildImagePipelineArtifactsBucketEncryptionKey5E194B47",
                "Arn",
              ],
            },
            "Type": "KMS",
          },
          "Location": {
            "Ref": "BuildImagePipelineArtifactsBucketC6CC33E0",
          },
          "Type": "S3",
        },
        "Name": "ubuntu_22_04BuildImagePipeline",
        "RestartExecutionOnUpdate": true,
        "RoleArn": {
          "Fn::GetAtt": [
            "BuildImagePipelineRole576404EF",
            "Arn",
          ],
        },
        "Stages": [
          {
            "Actions": [
              {
                "ActionTypeId": {
                  "Category": "Source",
                  "Owner": "AWS",
                  "Provider": "S3",
                  "Version": "1",
                },
                "Configuration": {
                  "S3Bucket": {
                    "Fn::ImportValue": "RepoStack:ExportsOutputRefBucket83908E7781C90AC0",
                  },
                  "S3ObjectKey": "data.zip",
                },
                "Name": "Build-Image-Source",
                "OutputArtifacts": [
                  {
                    "Name": "BuildImageSource",
                  },
                ],
                "RoleArn": {
                  "Fn::GetAtt": [
                    "BuildImagePipelineSourceBuildImageSourceCodePipelineActionRole30249AA3",
                    "Arn",
                  ],
                },
                "RunOrder": 1,
              },
            ],
            "Name": "Source",
          },
          {
            "Actions": [
              {
                "ActionTypeId": {
                  "Category": "Build",
                  "Owner": "AWS",
                  "Provider": "CodeBuild",
                  "Version": "1",
                },
                "Configuration": {
                  "ProjectName": {
                    "Ref": "BuildImageProjectD5B891B2",
                  },
                },
                "InputArtifacts": [
                  {
                    "Name": "BuildImageSource",
                  },
                ],
                "Name": "Build",
                "RoleArn": {
                  "Fn::GetAtt": [
                    "BuildImagePipelineBuildCodePipelineActionRole15E18A3A",
                    "Arn",
                  ],
                },
                "RunOrder": 1,
              },
            ],
            "Name": "Build",
          },
        ],
      },
      "Type": "AWS::CodePipeline::Pipeline",
    },
    "BuildImagePipelineArtifactsBucketC6CC33E0": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "KMSMasterKeyID": {
                  "Fn::GetAtt": [
                    "BuildImagePipelineArtifactsBucketEncryptionKey5E194B47",
                    "Arn",
                  ],
                },
                "SSEAlgorithm": "aws:kms",
              },
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "BuildImagePipelineArtifactsBucketEncryptionKey5E194B47": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::111111111111:root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Delete",
    },
    "BuildImagePipelineArtifactsBucketEncryptionKeyAlias74B7C629": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AliasName": "alias/codepipeline-repostackmyteststackbuildimagepipelineec25138e",
        "TargetKeyId": {
          "Fn::GetAtt": [
            "BuildImagePipelineArtifactsBucketEncryptionKey5E194B47",
            "Arn",
          ],
        },
      },
      "Type": "AWS::KMS::Alias",
      "UpdateReplacePolicy": "Delete",
    },
    "BuildImagePipelineArtifactsBucketPolicyA1FD352D": {
      "Properties": {
        "Bucket": {
          "Ref": "BuildImagePipelineArtifactsBucketC6CC33E0",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "BuildImagePipelineArtifactsBucketC6CC33E0",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "BuildImagePipelineArtifactsBucketC6CC33E0",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "BuildImagePipelineBuildCodePipelineActionRole15E18A3A": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::111111111111:root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "BuildImagePipelineBuildCodePipelineActionRoleDefaultPolicyF886D181": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "codebuild:BatchGetBuilds",
                "codebuild:StartBuild",
                "codebuild:StopBuild",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "BuildImageProjectD5B891B2",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "BuildImagePipelineBuildCodePipelineActionRoleDefaultPolicyF886D181",
        "Roles": [
          {
            "Ref": "BuildImagePipelineBuildCodePipelineActionRole15E18A3A",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "BuildImagePipelineEventsRole6FF38A6B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "events.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "BuildImagePipelineEventsRoleDefaultPolicy7384E26F": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "codepipeline:StartPipelineExecution",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":codepipeline:eu-central-1:111111111111:",
                    {
                      "Ref": "BuildImagePipeline01DF392E",
                    },
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "BuildImagePipelineEventsRoleDefaultPolicy7384E26F",
        "Roles": [
          {
            "Ref": "BuildImagePipelineEventsRole6FF38A6B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "BuildImagePipelineRole576404EF": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codepipeline.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "BuildImagePipelineRoleDefaultPolicy211D9F92": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "BuildImagePipelineArtifactsBucketC6CC33E0",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "BuildImagePipelineArtifactsBucketC6CC33E0",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "kms:Decrypt",
                "kms:DescribeKey",
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "BuildImagePipelineArtifactsBucketEncryptionKey5E194B47",
                  "Arn",
                ],
              },
            },
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "BuildImagePipelineSourceBuildImageSourceCodePipelineActionRole30249AA3",
                  "Arn",
                ],
              },
            },
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "BuildImagePipelineBuildCodePipelineActionRole15E18A3A",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "BuildImagePipelineRoleDefaultPolicy211D9F92",
        "Roles": [
          {
            "Ref": "BuildImagePipelineRole576404EF",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "BuildImagePipelineSourceBuildImageSourceCodePipelineActionRole30249AA3": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::111111111111:root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "BuildImagePipelineSourceBuildImageSourceCodePipelineActionRoleDefaultPolicy7FF981F0": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::ImportValue": "RepoStack:ExportsOutputFnGetAttBucket83908E77Arn063C8555",
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::ImportValue": "RepoStack:ExportsOutputFnGetAttBucket83908E77Arn063C8555",
                      },
                      "/data.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "BuildImagePipelineArtifactsBucketC6CC33E0",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "BuildImagePipelineArtifactsBucketC6CC33E0",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
                "kms:Decrypt",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "BuildImagePipelineArtifactsBucketEncryptionKey5E194B47",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "BuildImagePipelineSourceBuildImageSourceCodePipelineActionRoleDefaultPolicy7FF981F0",
        "Roles": [
          {
            "Ref": "BuildImagePipelineSourceBuildImageSourceCodePipelineActionRole30249AA3",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "BuildImageProjectD5B891B2": {
      "Properties": {
        "Artifacts": {
          "Type": "CODEPIPELINE",
        },
        "Cache": {
          "Type": "NO_CACHE",
        },
        "EncryptionKey": {
          "Fn::GetAtt": [
            "BuildImagePipelineArtifactsBucketEncryptionKey5E194B47",
            "Arn",
          ],
        },
        "Environment": {
          "ComputeType": "BUILD_GENERAL1_LARGE",
          "EnvironmentVariables": [
            {
              "Name": "ECR_REPOSITORY_URI",
              "Type": "PLAINTEXT",
              "Value": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::Select": [
                        4,
                        {
                          "Fn::Split": [
                            ":",
                            {
                              "Fn::ImportValue": "RepoStack:ExportsOutputFnGetAttRepository22E53BBDArn3AD4E30B",
                            },
                          ],
                        },
                      ],
                    },
                    ".dkr.ecr.",
                    {
                      "Fn::Select": [
                        3,
                        {
                          "Fn::Split": [
                            ":",
                            {
                              "Fn::ImportValue": "RepoStack:ExportsOutputFnGetAttRepository22E53BBDArn3AD4E30B",
                            },
                          ],
                        },
                      ],
                    },
                    ".",
                    {
                      "Ref": "AWS::URLSuffix",
                    },
                    "/",
                    {
                      "Fn::ImportValue": "RepoStack:ExportsOutputRefRepository22E53BBD9A9E013B",
                    },
                  ],
                ],
              },
            },
            {
              "Name": "AWS_ACCOUNT_ID",
              "Type": "PLAINTEXT",
              "Value": "111111111111",
            },
            {
              "Name": "AWS_DEFAULT_REGION",
              "Type": "PLAINTEXT",
              "Value": "eu-central-1",
            },
            {
              "Name": "IMAGE_TAG",
              "Type": "PLAINTEXT",
              "Value": "ubuntu_22_04",
            },
          ],
          "Image": "aws/codebuild/standard:7.0",
          "ImagePullCredentialsType": "CODEBUILD",
          "PrivilegedMode": true,
          "Type": "LINUX_CONTAINER",
        },
        "ServiceRole": {
          "Fn::GetAtt": [
            "BuildImageProjectRole3D242EB1",
            "Arn",
          ],
        },
        "Source": {
          "BuildSpec": "ubuntu_22_04/buildspec.yml",
          "Type": "CODEPIPELINE",
        },
      },
      "Type": "AWS::CodeBuild::Project",
    },
    "BuildImageProjectRole3D242EB1": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codebuild.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "BuildImageProjectRoleDefaultPolicyDE9DF999": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:eu-central-1:111111111111:log-group:/aws/codebuild/",
                      {
                        "Ref": "BuildImageProjectD5B891B2",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:eu-central-1:111111111111:log-group:/aws/codebuild/",
                      {
                        "Ref": "BuildImageProjectD5B891B2",
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "codebuild:CreateReportGroup",
                "codebuild:CreateReport",
                "codebuild:UpdateReport",
                "codebuild:BatchPutTestCases",
                "codebuild:BatchPutCodeCoverages",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":codebuild:eu-central-1:111111111111:report-group/",
                    {
                      "Ref": "BuildImageProjectD5B891B2",
                    },
                    "-*",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ecr:BatchCheckLayerAvailability",
                "ecr:GetDownloadUrlForLayer",
                "ecr:BatchGetImage",
                "ecr:CompleteLayerUpload",
                "ecr:UploadLayerPart",
                "ecr:InitiateLayerUpload",
                "ecr:PutImage",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::ImportValue": "RepoStack:ExportsOutputFnGetAttRepository22E53BBDArn3AD4E30B",
              },
            },
            {
              "Action": "ecr:GetAuthorizationToken",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "BuildImagePipelineArtifactsBucketC6CC33E0",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "BuildImagePipelineArtifactsBucketC6CC33E0",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "kms:Decrypt",
                "kms:DescribeKey",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "BuildImagePipelineArtifactsBucketEncryptionKey5E194B47",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "kms:Decrypt",
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "BuildImagePipelineArtifactsBucketEncryptionKey5E194B47",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "BuildImageProjectRoleDefaultPolicyDE9DF999",
        "Roles": [
          {
            "Ref": "BuildImageProjectRole3D242EB1",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "WeeklySchedule60EC10E3": {
      "Properties": {
        "ScheduleExpression": "cron(0 6 ? * Monday *)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition",
                  },
                  ":codepipeline:eu-central-1:111111111111:",
                  {
                    "Ref": "BuildImagePipeline01DF392E",
                  },
                ],
              ],
            },
            "Id": "Target0",
            "RoleArn": {
              "Fn::GetAtt": [
                "BuildImagePipelineEventsRole6FF38A6B",
                "Arn",
              ],
            },
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
